outlines/security.md

```diff
@@ -0,0 +1,55 @@
+# Security
+
+1. Main concept: Security surface area of a Meteor app
+    1. The way to security is to understand the attack surface, and guard against all possible attacks
+    2. Conclusion: Never Trust The Client - anyone can call all of your app's endpoints, not just the client you wrote. There's no way to guard against this, so you may as well not even try. Consider projects like SnapchatFS
+    3. It's not always clear where the server ends and the client begins in a Meteor app, but it's important to be aware of that for security
+2. Methods
+    1. Don't write generic methods, make sure you know what each argument is exactly, and what it could do
+    2. Rate limiting as a first line of defense against brute force
+    3. Make sure method side effects don't give away information. For example, returning how many items were affected should only tell the user about documents they should be able to access
+    4. Use this.userId, never take the current user as an argument in a method
+    5. Always include the user ID and any prerequisites directly in the query, in case that stuff changed since you did the security check
+    6. Check types, use audit-argument-checks and check-checker
+3. Publications
+    1. Most of the points about methods are still relevant!
+    2. Publications re-run when the user ID changes (describe this in detail)
+    3. Don't let people pass arbitrary queries into publications, this is bad both for performance and security
+    4. Make sure to filter the data so that the client can't get anything sensitive
+4. Served files
+    1. Your source code can contain secret information. If you have code with secret stuff in it:
+        1. Consider putting the secrets in settings
+        2. If that's not possible, make sure the file is in private/ or server/ so that it's not exposed to the client
+        3. Any settings files should be in private or a dotted folder like .config? This saves you from a poorly-coded .json build plugin exposing your file
+5. Secret keys in settings.json
+    1. Have a different settings file for each environment
+    2. Don't keep them in source control? If not, then where? LastPass?
```

Comment:
I actually don't know enough about this. Perhaps @multilinear knows the answer?

Brewer, where is the best place to keep secret API keys and similar for your app?